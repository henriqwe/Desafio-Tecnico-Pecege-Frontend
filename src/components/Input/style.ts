import styled from 'styled-components'

export const ErrorMessage = styled.label`
  display: block;
  font-size: 14px;
  padding-left: 0.5rem;
  font-weight: 600;
  color: #c01e2e;
  padding-top: 0.1rem;
`
export const InputWrapper = styled.div`
  display: block;
  flex: 1;
`

export const Container = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
`

export const Input = styled.input<{ isinvalid: 1 | 0; label: string }>`
  width: 100%;
  padding: 7px;
  padding-left: 18px;
  border: 2px solid ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#dadce0')};
  border-radius: 4px;
  font-size: 1rem;
  outline-color: transparent;
  &:focus {
    border: 2px solid ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#18a2f3')};
    outline-color: ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#18a2f3')};
  }
  + span::after {
    content: '${({ label }) => label}';
    display: flex;
  }
  &:not(:placeholder-shown) + span::after {
    transform: translateY(-6px);
  }
  &:focus + span::after {
    transform: translateY(-6px);
  }
  &:not(:placeholder-shown) + span {
    color: ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#18a2f3')};
    transform: translateX(10px) translateY(-17px);
    font-size: 0.75rem;
    font-weight: 600;
    background-color: #fff;
    height: 5px;
    padding: 0 6px;
  }

  &:focus + span {
    color: ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#18a2f3')};
    transform: translateX(10px) translateY(-17px);
    font-size: 0.75rem;
    font-weight: 600;
    background-color: white;
    height: 5px;
    padding: 0 6px;
  }

  &:not(:focus) + span {
    color: ${({ isinvalid }) => (isinvalid ? '#C01E2E' : '#808080')};
  }
`

export const Label = styled.span`
  position: absolute;
  left: 0;
  padding-left: 1.2rem;
  font-size: 1rem;
  color: #7f8fa6;
  pointer-events: none;
  transition: all 70ms ease-out;
`
