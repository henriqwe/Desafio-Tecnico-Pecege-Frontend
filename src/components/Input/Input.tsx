import {
  Control,
  Controller,
  DeepMap,
  FieldError,
  FieldValues,
  Path
} from 'react-hook-form'
import { useDebounce } from '&hooks/useDebounce'
import * as S from '&components/Input/style'

type InputType = {
  name: Path<FieldValues>
  control: Control<any, any>
  label: string
  handleChangeDebounce?: (value: string) => any
  debounceDelay?: number
  error?: DeepMap<FieldValues, FieldError> & { message?: string }
  disabled?: boolean
}
export function Input({
  name,
  control,
  handleChangeDebounce,
  error,
  label,
  debounceDelay = 1000,
  disabled = false
}: InputType) {
  const debounceChange = handleChangeDebounce
    ? useDebounce(handleChangeDebounce, debounceDelay)
    : undefined

  return (
    <Controller
      name={name}
      control={control}
      render={({ field: { onChange, value } }) => (
        <S.InputWrapper>
          <S.Label htmlFor={name}>{label}</S.Label>
          <S.Input
            autoComplete="off"
            id={name}
            type="text"
            disabled={disabled}
            value={value || ''}
            onChange={(e) => {
              onChange(e.target.value)
              debounceChange?.(e.target.value)
            }}
          />
          <S.ErrorMessage>{error?.message as string}</S.ErrorMessage>
        </S.InputWrapper>
      )}
    />
  )
}
