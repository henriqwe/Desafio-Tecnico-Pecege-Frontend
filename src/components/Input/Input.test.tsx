import { fireEvent, render, renderHook, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Input } from './Input'
import { useForm } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import * as Yup from 'yup'

export const schema = Yup.object().shape({
  name: Yup.string().required('Campo obrigatÃ³rio')
})

const onSubmitForm = jest.fn()
describe('Input', () => {
  it('should render correctly snapshot', async () => {
    const { result } = renderHook(() => useForm())

    const { container } = render(
      <Input control={result.current.control} label="teste" name="john due" />
    )

    expect(container).toMatchSnapshot()
  })
  it('not should submit when field is not filled ', async () => {
    const { result } = renderHook(() =>
      useForm({ resolver: yupResolver(schema) })
    )
    const {
      control,
      handleSubmit,
      formState: { errors }
    } = result.current

    const { findByTestId } = render(
      <form onSubmit={handleSubmit(onSubmitForm)}>
        <Input
          control={control}
          label="name"
          name="name"
          data-testid={'inputName'}
          error={errors['name']}
          value={undefined}
        />
        <button data-testid={'buttonOK'} type="submit">
          ok
        </button>
      </form>
    )

    const inputName = await findByTestId('inputName')

    const buttonOK = await findByTestId('buttonOK')

    await waitFor(() => fireEvent.click(buttonOK))
    expect(onSubmitForm).not.toBeCalled()
    fireEvent.change(inputName, { target: { value: 'Clementine Bauch Maria' } })

    await waitFor(() => fireEvent.click(buttonOK))

    expect(onSubmitForm).toBeCalledTimes(1)
  })
})
