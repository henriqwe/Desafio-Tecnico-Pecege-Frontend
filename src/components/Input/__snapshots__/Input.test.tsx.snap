// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly snapshot 1`] = `
<div>
  <div>
    <div
      class="sc-dkjKgF gJxtqW"
    >
      <input
        autocomplete="off"
        class="sc-hLclGa gyuuaE"
        id="john due"
        isinvalid="0"
        label="teste"
        placeholder=""
        type="text"
        value=""
      />
      <span
        class="sc-eDnVMP fkBieA"
      >
         
      </span>
    </div>
    <label
      class="sc-bcPKhP hSXdPs"
    />
  </div>
</div>
`;
